lib/travis/worker/virtual_machine/open_stack.rb

```diff
@@ -0,0 +1,270 @@
+require 'fog'
+require 'shellwords'
+require 'securerandom'
+require 'benchmark'
+require 'travis/support'
+require 'travis/worker/ssh/session'
+require 'resolv'
+require 'travis/worker/virtual_machine/blue_box/template'
+
+module Travis
+  module Worker
+    module VirtualMachine
+      class OpenStack
+        include Retryable
+        include Logging
+
+        DUPLICATE_MATCH = /testing-(\w*-?\w+-?\d*-?\d*-\d+-\w+-\d+)-(\d+)/
+
+        DEFAULT_TEMPLATE_ID = Travis::Worker.config.open_stack.default_template_id
+
+        USER_NAME = Travis::Worker.config.open_stack.template_username
+
+        class << self
+          def vm_count
+            Travis::Worker.config.vms.count
+          end
+
+          def vm_names
+            vm_count.times.map { |num| "#{Travis::Worker.config.vms.name_prefix}-#{num + 1}" }
+          end
+        end
+
+        log_header { "#{name}:worker:virtual_machine:open_stack" }
+
+        attr_reader :name, :password, :server, :ip_address
+
+        def initialize(name)
+          @name = name
+        end
+
+        # create a connection
+        def connection
+          @connection ||= Fog::Compute.new(
+            provider: :openstack,
+            openstack_api_key:  Travis::Worker.config.open_stack.api_key,
+            openstack_username: Travis::Worker.config.open_stack.username,
+            openstack_auth_url: Travis::Worker.config.open_stack.auth_url,
+            openstack_tenant:   Travis::Worker.config.open_stack.tenant,
+            connection_options: Travis::Worker.config.open_stack.connection_options || {}
+          )
+        end
+
+        def create_server(opts = {})
+          hostname = hostname(opts[:job_id])
+
+          config = open_stack_vm_defaults.merge(opts.merge({
+            :image_ref => template_id(opts),
+            :name => hostname
+          }))
+
+          retryable(tries: 3, sleep: 5) do
+            destroy_duplicate_servers
+            create_new_server(config)
+          end
+        end
+
+        def create_new_server(opts)
+          @password = (opts[:password] = generate_password)
+
+          opts[:user_data] = user_data(opts[:name], USER_NAME, opts[:password])

```

Comment:
what about windows?
It seams we will not use cloud-init because it really slows down start-up.
We will have pre-generated ssh keys for log-in.